src/backend/vulkan/src/pool.rs

```diff
@@ -0,0 +1,133 @@
+                }
+
+                let buffer = &mut self.command_buffers[self.next_buffer];
+                self.next_buffer += 1;
+
+                let info = vk::CommandBufferBeginInfo {
+                    s_type: vk::StructureType::CommandBufferBeginInfo,
+                    p_next: ptr::null(),
+                    flags: vk::COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT,
+                    p_inheritance_info: ptr::null(),
+                };
+
+                unsafe {
+                    self.device.0.begin_command_buffer(buffer.0.raw, &info); // TODO: error handling
+                    Encoder::new(buffer)
+                }
+            }
+
+            fn reset(&mut self) {
+                self.next_buffer = 0;
+                unsafe {
+                    self.device.0.fp_v1_0().reset_command_pool(
+                        self.device.0.handle(),
+                        self.pool,
+                        vk::CommandPoolResetFlags::empty()
+                    );
+                }
+            }
+
+            fn reserve(&mut self, additional: usize) {
+                unimplemented!()
+            }
+        }
+
+        impl pool::$pool for $pool {
+            fn from_queue<Q>(queue: &mut Q, capacity: usize) -> $pool
+                where Q: Into<$queue<CommandQueue>> + DerefMut<Target=CommandQueue>
+            {
+                // Create command pool
+                let info = vk::CommandPoolCreateInfo {
+                    s_type: vk::StructureType::CommandPoolCreateInfo,
+                    p_next: ptr::null(),
+                    flags: vk::CommandPoolCreateFlags::empty(),
+                    queue_family_index: queue.family_index,
+                };
+
+                let command_pool = unsafe {
+                    queue.device.0.create_command_pool(&info, None)
+                                .expect("Error on command pool creation") // TODO: better error handling
+                };
+
+                // Allocate initial command buffers
+                let info = vk::CommandBufferAllocateInfo {
+                    s_type: vk::StructureType::CommandBufferAllocateInfo,
+                    p_next: ptr::null(),
+                    command_pool: command_pool,
+                    level: vk::CommandBufferLevel::Primary,
+                    command_buffer_count: capacity as u32,
+                };
+
+                let command_buffers = unsafe {
+                    queue.device.0.allocate_command_buffers(&info)
+                                  .expect("Error on command buffer allocation") // TODO: better error handling
+                };
+                let command_buffers = command_buffers.into_iter().map(|buffer| {
+                    $buffer(
+                        CommandBuffer {
+                            raw: buffer,
+                            device: queue.device.clone(),
+                        }
+                    )
+                }).collect::<Vec<_>>();
```

Comment:
nit: I don't think you need to explicitly specify the collect type here, since it's inferred.